daemon/containerd/image_snapshot_unix.go

```diff
@@ -85,7 +95,50 @@ func (i *ImageService) unremapRootFS(ctx context.Context, mounts []mount.Mount)
 				return err
 			}
 
-			return os.Lchown(path, uid, gid)
+			return chownWithCaps(path, uid, gid)
 		})
 	})
 }
+
+// chownWithCaps will chown path and preserve the extended attributes.
+func chownWithCaps(path string, uid int, gid int) error {
+	xattrKeys, err := sysx.LListxattr(path)
+	if err != nil {
+		if !errors.Is(err, syscall.EOPNOTSUPP) {
+			return err
+		}
+	}
+
+	xattrs := make(map[string][]byte, len(xattrKeys))
+
+	for _, xattr := range xattrKeys {
+		data, err := sysx.LGetxattr(path, xattr)
+		if err != nil {
+			if !errors.Is(err, syscall.EOPNOTSUPP) {
+				return err
+			}
+			break
```

Comment:
Curious if we must break here, or continue with the next attribute; is it possible _some_ extended attributes are not supported, and we would now be skipping the remaining ones?

I also recall this PR from @corhere where we were silently swallowing errors, and wondering if discarding attributes is the right thing to do;

- https://github.com/moby/moby/pull/45464

Also looking at this existing code, which is not ignoring these; https://github.com/moby/moby/blob/0751141003ebc8fbacf3e30c453669e9ae8da2e5/daemon/graphdriver/copy/copy.go#L90-L106

Reply:
The `EOPNOTSUPP` errors is returned if `Extended attributes are not supported by the filesystem, or are disabled.` so if one fails, all will fail, that's why I break. The `copyXattr` _is_ ingoring it, it returns `nil` if it can't copy the xattrs.

Regarding Cory's PR, Cory is right, we should fail. But his PR doesn't cover all the cases, as you can see with the copyXattr case you just linked. If I fail here, do we want an escape hatch like the one in Cory's PR?

Reply:
> The EOPNOTSUPP errors is returned if Extended attributes are not supported by the filesystem, or are disabled. so if one fails, all will fail,

Gotcha! It wasn't immediately clear to me if this would "fail for _some_ (but not all)" or always for _all_.

> If I fail here, do we want an escape hatch like the one in Cory's PR?

As the containerd image store is "new territory", and assuming this would only happen in a "broken" setup (unsupported filesystem, so not "because user-NS"), I suggest that we should _not_ provide an escape-hatch out of the box. Worst case, we add one later if there's many complaints, but (as that escape hatch communicates through its ridiculous `i_want_broken_containers` name) we really shouldn't.

w.r.t. "xattrs are supported", I'm wondering if this is also information that snapshotters should provide and/or if they should somehow detect this and mark themselves as "not supported". I guess some of that may be more complicated with containerd being less opinionated (storage location and such, which may mean "any location could be used?")

Based on the above;

- I think we should probably bail out if it's not supported (return an error that extended attributes are not supported).
- If we _do_ want to "continue", at least it should be logged, and we should have a comment in code describing the intent "if the first one fails.. all of them will fail; it's not supported). So that it's clear that the intent is not to "skip some".